pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/user"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/pulumi/pulumi/pkg/v2/backend"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/diag"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/fsutil"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/workspace"
+)
+
+// PulumiFilestateLockingEnvVar is an env var that must be truthy to enable locking when using a filestate backend.
+const PulumiFilestateLockingEnvVar = "PULUMI_SELF_MANAGED_STATE_LOCKING"
+
+type lockContent struct {
+	Pid       int       `json:"pid"`
+	Username  string    `json:"username"`
+	Hostname  string    `json:"hostname"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func newLockContent() (*lockContent, error) {
+	u, err := user.Current()
+	if err != nil {
+		return nil, err
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	return &lockContent{
+		Pid:       os.Getpid(),
+		Username:  u.Username,
+		Hostname:  hostname,
+		Timestamp: time.Now(),
+	}, nil
+}
+
+// checkForLock looks for any existing locks for this stack, and returns a helpful diagnostic if there is one.
+func (b *localBackend) checkForLock(ctxt context.Context, stackRef backend.StackReference) error {
+	allFiles, err := listBucket(b.bucket, b.lockDir())
```

Comment:
Since the lock occurs against a particular stack, could we actually check by `b.lockPrefix(stack)` instead of the `b.lockDir()` and have `lockPrefix` return a bucket-style prefix ending with a `/`? I think that would make this more efficient?

Reply:
Good point - I believe you are right that we can enumerate just the locks for the given stacks.  I've made this change in https://github.com/pulumi/pulumi/pull/6437/commits/bc2273212865b81760c0dbe1545af80639580487.